### PR TITLE
Git fetcher: Calculate a fingerprint for dirty workdirs

### PR DESCRIPTION
**Pull request by edolstra**
## Motivation

This restores evaluation caching for dirty Git workdirs, which should help with #11985.

## Context

---

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libfetchers/git.cc [hunk 1]
```diff
@@ -793,10 +793,34 @@ struct GitInputScheme : InputScheme
...
-            return rev->gitRev() + (getSubmodulesAttr(input) ? ";s" : "") + (getExportIgnoreAttr(input) ? ";e" : "");
-        else
+            return makeFingerprint(*rev);
+        else {
+            auto repoInfo = getRepoInfo(input);
+            if (repoInfo.isLocal && repoInfo.workdirInfo.headRev) {
+                /* Calculate a fingerprint that takes into account the
+                   deleted and modified/added files. */
+                HashSink hashSink{HashAlgorithm::SHA512};
+                for (auto & file : repoInfo.workdirInfo.files)
```
Comment (Mic92) [inline src/libfetchers/git.cc:809]: This ignores untracked files, right? People sometimes unpack large directories not tracked by git into their projects.

Comment (edolstra) [reply src/libfetchers/git.cc:809]: That's right, but untracked files are invisible to the evaluator so it's correct to not include them in the fingerprint.

**Review by roberth — COMMENTED** [review]
I'm not opposed to this as an incremental improvement, but this could likely be made faster, simpler and we could fix one or two bugs along the way.

- `WorkdirInfo` fundamentally requires that we do the equivalent of a `git status`, which reads the whole workdir.
- It should be quick to produce a commit from a `git_status_list` or the intermediate data that produced it
- We ought to `clean` the files before using them, just as the remotely fetched blobs aren't smudged
- Since the mempack PR, we have infrastructure to create objects in memory. By creating a tree in memory, we can use the same code for workdir and commit-based fetching. This significantly reduces complexity in the git fetcher. (and solves silently broken cache entries when files are mutated concurrently; should be rare)

**Comment by jsqu4re** [comment]
Thank you so much for tackling this! That's exceptional responsive :+1: 

I am not sure whether this suggestions fits into this PR:

We are currently using a _self-referencing flake_ on the top level of a mono repo with **sparse-checkout**, where everything missing is found using the self-reference as input of the flake. The sparse mono repo checkout is only a couple of files and the developer checkouts the folders he wants to work with. This makes flakes in mono repos quite quick. The one thing that would be needed to do this a bit better is either supporting sparse-checkout (with [libgit2 not supporting it](https://github.com/libgit2/libgit2/issues/2263)) or ignoring the changes coming from missing files due to sparse checkout. Which might not be nice but enables the use case, or a third option I am not aware of :stuck_out_tongue_winking_eye:

**Comment by jsqu4re** [comment]
``` shell
[nixos@nixos:~/testing/flake-small]$ nix --version
nix (Nix) 2.24.10

[nixos@nixos:~/testing/flake-small]$ time nix build .#packages.x86_64-linux.hello
warning: Git tree '/home/nixos/testing/flake-small' is dirty

real    0m0.564s
user    0m0.339s
sys     0m0.127s

[nixos@nixos:~/testing/flake-small]$ nix shell github:DeterminateSystems/nix-src/a799ea4368bcd1ad31e75afbe095acebf8df0ed3

[nixos@nixos:~/testing/flake-small]$ nix --version
nix (Nix) 2.26.0pre

[nixos@nixos:~/testing/flake-small]$ time nix build .#packages.x86_64-linux.hello
warning: Git tree '/home/nixos/testing/flake-small' is dirty

real    0m0.522s
user    0m0.355s
sys     0m0.082s

[nixos@nixos:~/testing/flake-small]$ time nix build .#packages.x86_64-linux.hello
warning: Git tree '/home/nixos/testing/flake-small' is dirty

real    0m0.523s
user    0m0.355s
sys     0m0.081s

[nixos@nixos:~/testing/flake-small]$ time nix build .#packages.x86_64-linux.hello
warning: Git tree '/home/nixos/testing/flake-small' is dirty

real    0m0.106s
user    0m0.073s
sys     0m0.015s
```

**Comment by jsqu4re** [comment]
For a bigger example it seems to be slower than before
nix 2.24 = 0.7s 
to this PR = 1.3s down to 1.0s
This for the example with sparse-checkout

**Comment by roberth** [comment]
For context, libgit2 status code can still be optimized significantly
- https://github.com/libgit2/libgit2/issues/5038

**Comment by edolstra** [comment]
I've added some caching that speeds up `nix flake metadata` on a dirty Nixpkgs repo from 2.8s to 0.7s.

**Comment by edolstra** [comment]
> For context, libgit2 status code can still be optimized significantly (and proven)

Indeed, `nix flake metadata` on a nixpkgs workdir spends about 66% of its time in `git_status_foreach_ext()` (which is about 0.5s, compared to 0.2 for `git status` on the same repo).

**Comment by jsqu4re** [comment]
With these changes my example is down to 400 ms

**Comment by edolstra** [comment]
> WorkdirInfo fundamentally requires that we do the equivalent of a git status, which reads the whole workdir.

That's pretty unavoidable though. You have to do a full traversal of the workdir to see if any files have been modified.

Not sure if creating a commit really simplifies things. Workdir support is really just a call to `FSSourceAccessor`, which we need anyway.

**Comment by roberth** [comment]
> > status

> unavoidable

:+1: 

> Not sure if creating a commit really simplifies things. Workdir support is really just a call to `FSSourceAccessor`, which we need anyway.

Unfortunately `FSSourceAccessor` is not a complete solution because we don't `clean` (un-`smudge`) the files yet.

The amount of duplication between workdir and tree-based code paths is non-trivial and any tiny divergence between those will end up in a hash mismatch for someone.

File: src/libfetchers/git.cc [hunk 2]
```diff
@@ -793,10 +794,33 @@ struct GitInputScheme : InputScheme
...
+        else {
+            auto repoInfo = getRepoInfo(input);
+            if (repoInfo.isLocal && repoInfo.workdirInfo.headRev && repoInfo.workdirInfo.submodules.empty()) {
+                /* Calculate a fingerprint that takes into account the
+                   deleted and modified/added files. */
+                HashSink hashSink{HashAlgorithm::SHA512};
+                for (auto & file : repoInfo.workdirInfo.dirtyFiles) {
+                    writeString("modified:", hashSink);
+                    writeString(file.abs(), hashSink);
+                    dumpPath(repoInfo.url + "/" + file.abs(), hashSink);
```
Comment (roberth) [inline src/libfetchers/git.cc:813]: How are symlinks and the executable bit handled?

**Review by roberth — CHANGES_REQUESTED** [review]
I have no confidence in the correctness of this approach, but it might be fast

File: src/libfetchers/git.cc (hunk 2 above)
Comment (edolstra) [reply src/libfetchers/git.cc:813]: `dumpPath()` dumps a NAR, so it handles symlinks and the executable bit.

**Comment by edolstra** [comment]
> The amount of duplication between workdir and tree-based code paths is non-trivial and any tiny divergence between those will end up in a hash mismatch for someone.

That may be true, but it's orthogonal to this PR since we're already using `FSSourceAccessor` for dirty workdirs.

**Review by roberth — APPROVED** [review]
From work meeting:
Add test case for `nix flake metadata`, check that fingerprint exists in dirty worktree.

@roberth: Approved. Performance overrides the tech debt concern. We'll get to it.

**Comment by Mic92** [comment]
@mergify queue

**Comment by kjeremy** [comment]
Can we get this backported?

**Pull request merged**